dnup code review

dsplaisted opened the pull request:
We're not planning to merge this to main yet, but we want to start reviewing the code.

Context: https://github.com/dotnet/designs/pull/339

dsplaisted committed:
Initial dotnet sdk install CLI definition

dsplaisted committed:
Add SDK update command

dsplaisted committed:
Rename VersionOrChannel argument to just Channel

dsplaisted committed:
Implement part of UI for install command

dsplaisted committed:
Add more installer UI

dsplaisted committed:
UI improvements and add dnup shims for demo

dsplaisted committed:
More install experience updates

dsplaisted committed:
Add SdkInstallCommand code to dnup

dsplaisted committed:
Wire up sdk install command for dnup

dsplaisted committed:
Add string resources to dnup project

dsplaisted committed:
Add initial installer interfaces for dnup

dsplaisted committed:
Add initial installer interfaces for dnup (#50379)

dsplaisted committed:
Start implementation of DotnetInstaller

Copilot prompt:

Implement the GetConfiguredInstallType method in #file:'DotnetInstaler.cs' .

This method should look in the PATH and resolve "dotnet" or "dotnet.exe", depending on the current OS.  If it is not found, the return value should be SdkInstallType.None.  If it is found in Program Files, the install type should be SdkInstallType.Admin.  If it is found in another folder, the install type should be SdkInsntallType.User.

However, the method should also check the value of DOTNET_ROOT environment variable.  For a user install, DOTNET_ROOT should be set to the folder where the dotnet executable is found.  For an admin install, DOTNET_ROOT should not be set, but if it is set to the install path under program files, that is OK.  If the DOTNET_ROOT value doesn't match the install location, the method should return SdkInstallType.Inconsistent.

dsplaisted committed:
Add dnup solution filter and add Environment provider to project

dsplaisted committed:
Switch to common implementation of finding dotnet on path

Copilot prompt:

In #method:'Microsoft.DotNet.Tools.Bootstrapper.DotnetInstaler.GetConfiguredInstallType':301-2793 ,  use #method:'Microsoft.DotNet.Cli.Utils.EnvironmentProvider.GetCommandPath':2399-2843 to search for dotnet on the path.

dsplaisted committed:
Fix typo

dsplaisted committed:
Add class for global.json contents

Copilot prompts:

Create a class to represent the contents of a global.json file, and create a corresponding JsonSerializerContext type.

Can you refer to https://learn.microsoft.com/en-us/dotnet/core/tools/global-json to update the schema you are using for the global.json file?  AllowPrerelease should be a booleann, and instead of Path there should be a Paths property which is an array of strings.

dsplaisted committed:
Rename class and more

dsplaisted committed:
Implement GetGlobalJsonInfo

Copilot prompts:

Implement #method:'Microsoft.DotNet.Tools.Bootstrapper.DotnetInstaller.GetDefaultDotnetInstallPath':2814-2986 .  It look for a global.json file starting in the specified directory and walking up the directory chain.  If it finds one, it should deserialize its contents using System.Text.Json APIs and #class:'Microsoft.DotNet.Tools.Bootstrapper.GlobalJsonContentsJsonContext':413-650 , and use that for the return value.

Sorry, I meant to ask you to implement #method:'Microsoft.DotNet.Tools.Bootstrapper.DotnetInstaller.GetGlobalJsonInfo':4307-4411  instead of #method:'Microsoft.DotNet.Tools.Bootstrapper.DotnetInstaller.GetDefaultDotnetInstallPath':2839-2947 .  Can you revert your changes and then implement #method:'Microsoft.DotNet.Tools.Bootstrapper.DotnetInstaller.GetGlobalJsonInfo':4307-4411  using the instructions I previously provided?

dsplaisted committed:
Remove try/catch for loading global.json

Copilot prompts:

Can you remove the try catch block around reading the global.json file in #method:'Microsoft.DotNet.Tools.Bootstrapper.DotnetInstaller.GetGlobalJsonInfo':3019-4216 ?  Also remove the convenience properties you added to GlobalJsonInfo for now.

OK, go ahead and add those convencience properties back.

dsplaisted committed:
Implement GetConfiguredInstallType and GetGlobalJsonInfo (#50408)

dsplaisted committed:
Add more installer methods

dsplaisted committed:
Add calls to new methods

dsplaisted committed:
Initial ConfigureInstallType implementation

Copilot prompt:

Implement the #method:'Microsoft.DotNet.Tools.Bootstrapper.DotnetInstaller.ConfigureInstallType':4479-4606  method.  If the install type is user, remove any other folder with dotnet in it from the PATH and add the dotnetRoot to the PATH.  Also set the DOTNET_ROOT environment variable to dotnetRoot.

If the install type is Admin, unset DOTNET_ROOT, and add dotnetRoot to the path (removing any other dotnet folder from the path).

If the install type is None, unset DOTNET_ROOT and remove any dotnet folder from the path.

For any other install type, throw an ArgumentException.

dsplaisted committed:
Fix check for dotnet folder in PATH

Copilot prompt:

In #method:'Microsoft.DotNet.Tools.Bootstrapper.DotnetInstaller.ConfigureInstallType':4481-6459 what I meant by removing a folder from the path if it has dotnet in it was that you should check the contents of each folder and if it is a dotnet installation folder then it should be removed from the path.  A simple way to check if it's a dotnet installation folder is if it has a dotnet executable in it.

dsplaisted committed:
Add more methods to installer interface (#50475)

nagilson committed:
Add boilerplate types and interfaces to do installs

I renamed the InstallType as the Runtime installs would similarly hold the same properties.

nagilson committed:
Add base version class

nagilson committed:
Add dotnetVersion class for version parsing

nagilson committed:
Add dotnetVersion class for version parsing (#50492)

nagilson committed:
Add basic tests

nagilson committed:
Add basic tests for dnup (#50497)

nagilson committed:
Replace SDKInstallType with InstallType

nagilson committed:
Add Base Interface for Install

I renamed the DotnetInstaller in an aim to refactor the logic out some more. I think we could have the controller logic that talks to the CLI / UI layer be separate from the class that does the actual install logic to better follow the single responsibility principle and simplify the code into more pieces.

nagilson committed:
Add Base Interface for Install (#50498)

nagilson committed:
Add the isolated classes for each responsibility of install

nagilson committed:
Add the isolated classes for each responsibility of install (#50500)

nagilson committed:
fill in controller with correct data structure model

nagilson committed:
use some actual version parsing

nagilson committed:
add more context into what to implement for future me or others

nagilson committed:
Add Interfaces for .NET Installer Data Models (#50503)

nagilson committed:
prepare code from dnvm

make sure to properly give credit in the real thing, we may want to avoid ifilesystem, but not sure yet.

nagilson committed:
Prepare some install logic from dnup  (#50504)

nagilson committed:
Migrate to .NET Archive Libraries

We should try to use our default libraries when possible instead of shelling out to the tar process. There were some issues with the library with null terminators, but they have been resolved in .net 10 preview 4.

nagilson committed:
Implement Extraction Logic

nagilson committed:
Implement DNUP Extraction logic (#50626)

nagilson committed:
Implement manifest parsing and vscode setting folder for build button

nagilson committed:
Implement manifest parsing  (#50642)

nagilson committed:
Installing .NET works e2e

nagilson committed:
Installing .NET works e2e with dnup (#50645)

nagilson committed:
Show progress correctly

nagilson committed:
fix end message

nagilson committed:
find available SDKs

nagilson committed:
Working Progress indicators (#50647)

nagilson committed:
Add update cadence method for manifest tracking

the objects should be public to be json-ifiable

nagilson committed:
Add implementation of manifest methods

nagilson committed:
Extract + Install combined to one step

nagilson committed:
Fix json aot serialization

nagilson committed:
find existing versions + parse channels

nagilson committed:
find existing versions

nagilson committed:
Add tests for version parsing.

nagilson committed:
Correctly parse 9.0.1xx, and 9, and 9.0

nagilson committed:
Fix version parsing

nagilson committed:
add lts sts schannel support. fix project.

nagilson committed:
add preview support.

nagilson committed:
Add dnup manifest + channel parsing (#50665)

nagilson committed:
Fix LTS STS Parsing

nagilson committed:
Fix LTS STS Parsing (#50688)

dsplaisted committed:
Add dnup tests to dnup.slnf

dsplaisted committed:
Move sdk update command to dnup

dsplaisted committed:
Misc cleanup

dsplaisted committed:
Cleanup

dsplaisted committed:
Remove dnup to dotnet shims

nagilson commented:
> Higher level comments:
> 
> Should the implementations of the installer / downloader be calling (Spectre) console APIs directly, or should that be abstracted out?
> 
> * DotnetInstall
>   
>   * InstallMode - Eventually I don't think we'll want a hard-coded enum here.  Eventually we'll probably want different components that may be optional and should probably be data driven.
>   * ManagementCadence - This type is confusing to me, I'm not really sure what it represents.  Perhaps the name could be improved.
>   * There's overlap between DotnetInstallRequest and DotnetInstall, which is especially confusing when both of them are passed to ArchiveDotnetInstaller.  Probably we can improve the factoring here.
>   * How do we represent a “channel”?  Is that with DotnetInstallRequest?
> * DotnetVersion - Can we remove this and use ReleaseVersion directly?  Otherwise we should make sure we're not duplicating code here from elsewhere.
> * ArchiveDotnetInstaller – A lot of logic seems to be duplicated between zip and .tar.gz.  Should create an abstraction over those to reduce duplicated code?
> * IBootstrapperController - I think this should be renamed, maybe to IDotnetInstaller
>   
>   * InstallerOrchestratorSingleton – can this functionality be merged in?  It can still use a singleton, but that could be behind the interface
>   * I see there's already an IDotnetInstaller interface.  I think maybe all of this functionality can be merged into one interface.  The implementation would still handle the singleton and Prepare / Commit, but you wouldn't need to do that for a test mock implementation
> * Dnup manifest - I think eventually we'll want a more complicated structure to this information.  I think we'll want to keep track of installations separately from "references" or "installation requests".  We might have one installation for multiple reasons, so we need to keep track of the references separately to do garbage collection and uninstallation.
> * ReleaseManifest
>   
>   * This seems like it may be a lot more complicated than it needs to be and have a lot of repeated code.  This also relates to how we represent versions and channels, ideally we can have one method that gets the latest matching version available for a channel or version.
>   * Does the ProductCollectionApi handle caching downloads in-memory?  If not it looks like we may end up downloading the manifests multiple times during one operation
> 
> Further TODO items:
> 
> * Support elevation to set registry key to stop admin installer from updating PATH, as well as removing dotnet from the admin PATH
> * Actually report progress of download
> * Translate global.json info into channel (right now this is just stubbed out)

Overview based on our conversation:

We could change InstallMode to be a class so we can componetize the SDK more, and support including the runtime and sdk into a single folder, for example.
I don't think we should get rid of it entirely as this is a regret from the old vscode extension design I had.

ManagementCadence describes whether an install is standalone, global.json managed, or dnup managed when running dnup update.

I agree we should prefer composition and aggregation over inheritance for DotnetInstall - the DotnetInstlall might as well have the chanel data too.

A channel is currently represented in the request.

I think we should remove dotnet version when applicable. We may need to keep the compare methods or a few other methods. 

ArchiveDI - This seems like a wise idea though lower priority.

IBootstrCtrlr - I don't think we need to have a singleton as long as the windows /unix mutex library doesn't let the same process re-grab the mutex (btw this would also be broken now.) I would prefer to keep the 3 layers of abstraction for a library that doesn't want to have a dnup manifest or other dnup functionality. 

DNUP Manifest - Yes, I agree we might want to ref count. Great point.

Release Manifest- Yes, let's simplify this. I think the DL progress is correct but the cache is supposed to be file system based, yet it is not ysed. This is mostly copilot.

Support Elevation - Yes, we should make a separate issue to keep track of that work to set the path manip and reg key. 
https://github.com/dotnet/sdk/issues/51124 Here it is!

nagilson commented on src/Installer/dnup/dnup.sln:
```diff
@@ -0,0 +1,24 @@
+# Visual Studio Version 17
+VisualStudioVersion = 17.5.2.0
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "dnup", "dnup.csproj", "{FE7DE2FC-400E-D3A1-8410-80BAA0C888BE}"
```
Yes, I agree.

dsplaisted committed:
Miscellaneous dnup cleanup (#51122)

nagilson commented on src/Installer/dnup/IDnupManifest.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    internal interface IDnupManifest
+    {
+        IEnumerable<DotnetInstall> GetInstalledVersions(IInstallationValidator? validator = null);
```
What if we want a validator that is simple (checks if dotnet.exe exists and the runtime or sdk is found running dotnet --list-sdks or dotnet-runtimes?

Then we could also have a full validator that knows every single file in the install and makes sure it is there.

For test, we could have a validator that always validates no matter if its on disk, so we can have tests manifests that use fake installs on disk that doesn't exist, and we dont want to fail if its not on disk.

dsplaisted committed:
Refactor DotnetInstall types

dsplaisted committed:
Add library for installer functionality

dsplaisted committed:
Fix json serialization after refactor

dsplaisted committed:
Add initial interfaces to installation library

dsplaisted committed:
Refactor DotnetInstall types (#51140)

dsplaisted committed:
Flesh out installer interfaces

dsplaisted committed:
Update install root, install discoverer, and install type APIs

dsplaisted committed:
Add interfaces for installation library (#51184)

nagilson committed:
Move the launching workspace into a root folder for the lib and dnup

nagilson committed:
Use a clear, empty terminal window for launching rerun

nagilson committed:
Add entire SDK repo to workspace so it's easier to work with both contexts.

nagilson committed:
wip - prepare for e2e test

nagilson committed:
Add copilot instructions to use powershell over cmd for dnup

nagilson committed:
Add --no-progress because spectre prevents concurrent interactive displays.

nagilson committed:
Delete extraneous test file

nagilson committed:
Add end to end tests for dnup

This is slow and a less mature approach to unit testing, although I think we should focus our efforts in integration and e2e testing for now to ensure the broad cases work. This code will need better clean up later and has some re-implemented logic.

nagilson committed:
Remove extra concurrency tests as the e2e ones are already parallel

nagilson committed:
Add End to End Dnup Tests (#51347)

nagilson committed:
Add basic yml for running PR tests in dnup

we will want to consolidate this into the SDK later.

run pr ci on dnup

I don't have permission to add a pipeline so let's try to filter them out

I wish there was a built in method for yml to do this instead. For example, pr: has exclude which lets us not run the ci if changes are only in certain components. Ideally we could do something similar for the stages and condition them based on dnup/ non dnup changes? Note that we still need to have the .vsts-pr.yml file as the root pipeline file, and we can't just exclude running the entire thing as it controls the SDK pipeline as well.

dsplaisted committed:
Remove DotnetVersion

dsplaisted committed:
Rename IBootstrapperController to IDotnetInstallManager

dsplaisted committed:
Delete one of the IDotnetInstaller interfaces

dsplaisted committed:
Move ArchiveDotnetInstaller and dependencies to library

nagilson committed:
Remove modifications to main CI

Working with the engineering system team to make a new pipeline in azdo since I cannot. adding this bloat to the main SDK pipeline is not the ideal approach.

nagilson committed:
Improve yml file for dnup test runs

nagilson committed:
Run `dnup` tests in the `dnup` branch pipeline if only dnup code changed (#51349)

nagilson committed:
Merge remote-tracking branch 'upstream/main' into dnup

nagilson committed:
Sync Main into `dnup` branch (#51370)

nagilson committed:
Fix formatting of pipeline

nagilson committed:
Fix yml whitespace

nagilson committed:
Try to use oneES Job Template

nagilson committed:
fix whitespace

nagilson committed:
fix whitespace

nagilson committed:
Fix parameters

nagilson committed:
Update test names and use proper arm64 test template

nagilson committed:
fix whitespace

nagilson committed:
Include helix target queue in name

nagilson committed:
fix merge

nagilson committed:
Try to use .dotnet local SDK for .NET 10 SDK

nagilson committed:
fix missing $ in yml

nagilson committed:
Use different separator for bootstrap step

nagilson committed:
Include EnvDetectionRule

nagilson committed:
Add unique identifier to build steps to make it easier to tell from afar

nagilson committed:
Use a separate icon to separate build steps

im not sure if I really like this - chose a fortress to represent 'guarding' the code by running tests.... is it really helpful? It might be if / when we have other test steps to differentiate and find the dnup test checks.

nagilson committed:
Don't use global/ in the mutex

there is no concept of a global\ mutex prefix on unix which could be incorrect. On windows, this makes it so the mutex is not for the local user session but instead across all users. I'm not sure if this is needed but since we edit the system path, it might be. I will have to think about whether there will be issues with a higher-level admin user that has held onto a mutex and got stuck, and whether a lower level user could undo that.

nagilson committed:
Wait for longer to acquire mutex

10 seconds may fail - concurrently the tests could take a while. lets not fail the test because other tests are running.

nagilson committed:
Preserve stacktrace with throw;

nagilson committed:
Make test artifacts dir so it doesn't fail to publish if tests fail

nagilson committed:
Prefer checking if the thread holds the mutex rather than OpenExisting

OpenExisting may run into challenges with security policies changing - let's not try to navigate that when we can track it ourselves.

nagilson committed:
Improve manifest tests

dsplaisted committed:
Create public APIs in library for installation

nagilson committed:
Add manifest path as an option

this would likely cause concurrency issues in tests if we just rely on the environment variable.

nagilson committed:
Fix expressions

nagilson committed:
don't consider 20s as a long test - installing in an e2e test takes a while

nagilson committed:
use custommanifest path in IRA check

nagilson committed:
increase mutex wait time before giving up

tests total runtime is 200 s on my machine - let's see if this helps. this would ideally be configurable because most users probably won't run multiple installs, though also they may have a much slower machine where this would cause concurrent requests to fail.

nagilson committed:
use a separate process so mutexes aren't shared improperly

nagilson committed:
remove console output capture - extraneous now

dsplaisted committed:
Add library API to get latest release in a channel

nagilson committed:
Manifest paths should include custom path

may refactor this later...

nagilson committed:
Add CWD

nagilson committed:
Add cwd to tests to try to find the process

nagilson committed:
find executable more correctly

we probably shouldnt hard code debug amongst other things. Im sure theres a better approach but just want to get it green first.

nagilson committed:
Merge branch 'dnup' into library-implementation

nagilson committed:
Add concurrency test + Installation Validation

This implements validation of installs using hostfxr apis to ensure the install actually works and not just that the manifest is tracking the installs correctly in e2e tests.

It also adds a test to show that we can do multiple installs in the same directory without failing.

It also improves the existing test logic to not assume a hard-coded debug value for the dnup process.

nagilson committed:
Add concurrency test + Installation Validation

This implements validation of installs using hostfxr apis to ensure the install actually works and not just that the manifest is tracking the installs correctly in e2e tests.

It also adds a test to show that we can do multiple installs in the same directory without failing.

It also improves the existing test logic to not assume a hard-coded debug value for the dnup process.

nagilson committed:
Revert "Add concurrency test + Installation Validation"

This reverts commit 0fc7172351862847ce4abbd9443eb47d1d2ed838.

nagilson committed:
Run tests for `dnup` branch across all 3 OS in PRs (#51372)

nagilson committed:
Merge remote-tracking branch 'upstream/dnup' into library-implementation

nagilson committed:
fix  strong name issue

nagilson committed:
Allow public use of GetInstallArchitecture

we can chat if we want this to be internal - though it might be helpful for consumers to be able to marshal the architecture from the runtime into our architecture type that we use in our install objects, considering it's part of dotnet install root, which is also public

dsplaisted committed:
Add basic ability to get latest version and install to installation library (#51374)

nagilson committed:
remove extra GCP fnct

nagilson committed:
Fix new references

nagilson committed:
Enable debugging into dnup process in test

nagilson committed:
Create a window if we are in debug mode so we can press enter.

nagilson committed:
Can properly attach to test debug process

though now it always expects an attacher

nagilson committed:
Switch to environment variable that enables debug launching into the test.

nagilson committed:
Workspace should enable debugging singular test

nagilson committed:
Fix Version Parsing issues in ReleaseManifest

nagilson committed:
Test does not need to hold onto mutex during validation

nagilson committed:
Simplify release manifest logic

nagilson committed:
Extensively simplify product version filtering logic

nagilson committed:
even further simplify the version parsing logic

nagilson committed:
Copy the logic for sdk CI to dnup CI

nagilson committed:
Produce a package from the library on build

nagilson committed:
enable package creation

nagilson committed:
Package version correctly created at D:\sdk\artifacts\packages\Release\NonShipping

nagilson committed:
Simplify package creation for now to only run win tests / pkg

nagilson committed:
remove unused publish properties

nagilson committed:
Address Feedback

nagilson committed:
add to dnup slnf

nagilson committed:
Validate .NET Installs from `dnup` (#51405)

nagilson committed:
Merge remote-tracking branch 'upstream/dnup' into nagilson-nuget-pipeline

nagilson committed:
`dnup` CI Pipeline (#51437)

jaredpar commented on src/Installer/dnup/DnupSharedManifest.cs:
```diff
@@ -0,0 +1,135 @@
+        AssertHasFinalizationMutex();
+        EnsureManifestExists();
+
+        var json = File.ReadAllText(ManifestPath);
```
This reads the entire content of the file into memory. Did you consider using a `Stream` for deserialization? Intuition is that is going to be more efficient.

nagilson replied:
That's a great idea. The manifest file will be very small, so I'm curious to see the numbers when we implement that. The manifest is very barebones - eventually we want a json schema file, caching, ability to upgrade / work with breaking changes to the manifest, manifest version metadata, etc. All of that is tracked in https://github.com/dotnet/sdk/issues/51099, and I'll add the `Stream` idea, thank you!

nagilson commented on src/Installer/dnup/DnupSharedManifest.cs:
```diff
@@ -0,0 +1,110 @@
+internal class DnupSharedManifest : IDnupManifest
+{
+    private static readonly string ManifestPath = Path.Combine(
+        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
```
Good call... I'll revisit the spec on default install locations and paths.

nagilson commented on src/Installer/dnup/DotnetInstall.cs:
```diff
@@ -0,0 +1,47 @@
+    InstallMode Mode,
+    InstallArchitecture Architecture)
+{
+    public Guid Id { get; } = Guid.NewGuid();
```
I think an Id would be helpful because multiple global json may depend on the same version/path but the install id would let us differentiate between those two 'installs' as dependents of the same install. I suppose its a different way of thinking about the installs.

nagilson commented on src/Installer/dnup/DnupSharedManifest.cs:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    private void AssertHasFinalizationMutex()
```
RAII patterns with Dispose are definitely preferred. I think this function should be replaced with logic that acquires the mutex every function call within the manifest to prevent this from being something othes need to worry about. I need to thoroughly review the C# mutex behavior but im pretty sure a process can reacquire if it already owns the mutex so this would be fine. This 'assert' was here because I very much want the program to crash  if there is ever a condition in which we modify the manifest without holding a mutex because that would indicate some code is fundamentally flawed - probably I would condition it to not be compiled in #RELEASE mode. Anyway, that's a C++ pattern I am using here.

nagilson committed:
Fix whitespace and package name

nagilson committed:
See if we need a separate package step or not, move into job

nagilson committed:
Don't publish any artifact, not allowed as build artifact

nagilson committed:
use pipeline artifact output

nagilson committed:
Create dnup library package

nagilson committed:
Create separate arcade build script package step

nagilson committed:
Add link to build def id

nagilson commented on src/Installer/dnup/BootstrapperController.cs:
```diff
@@ -0,0 +1,176 @@
+        string programFiles = Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles);
+        string programFilesX86 = Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86);
+        bool isAdminInstall = installDir.StartsWith(Path.Combine(programFiles, "dotnet"), StringComparison.OrdinalIgnoreCase)
+            || installDir.StartsWith(Path.Combine(programFilesX86, "dotnet"), StringComparison.OrdinalIgnoreCase);
```
Instead, I think we should have a more sophisticated way to tell if the install is an admin-based install that works across OS layers. The CLI caller that represents this doesn't seem to differentiate between whether this is a windows only flag or not. I've added a TODO to improve this logic. `dnup` for now doesn't modify the `PATH`.

nagilson committed:
Produce all packages and separate them based on arcade build

nagilson committed:
remove arcade logic - it looks like it's trying to sign without arcade, but the signing step fails because its not a release/* branch

nagilson committed:
No need to restore in CI

nagilson committed:
Move dnup templates into dnup folder

nagilson committed:
Disable SDK CI on dnup/release branch checks if only dnup code changes

need to confirm if we change sdk code that things still run for the sdk code.

nagilson committed:
add release/dnup

nagilson committed:
Produce `Dotnet.Installation` library package for internal testing in `dnup` ci (#51472)
